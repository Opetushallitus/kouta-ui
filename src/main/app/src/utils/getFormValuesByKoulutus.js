export const getFormValuesByKoulutus = koulutus => {
  const {
    kielivalinta = [],
    koulutusKoodiUri = '',
    koulutustyyppi = '',
    tarjoajat = [],
    metadata = {},
    nimi = {},
    julkinen = false,
    tila,
    teemakuva,
  } = koulutus;

  const {
    lisatiedot = [],
    kuvaus = {},
    opintojenLaajuusKoodiUri = '',
    tutkintonimikeKoodiUrit = [],
    kuvauksenNimi = {},
    koulutusalaKoodiUrit = [],
  } = metadata;

  const osiot = lisatiedot
    .filter(({ otsikkoKoodiUri }) => !!otsikkoKoodiUri)
    .map(({ otsikkoKoodiUri }) => ({ value: otsikkoKoodiUri }));

  const osioKuvaukset = lisatiedot.reduce((acc, curr) => {
    if (curr.otsikkoKoodiUri) {
      acc[curr.otsikkoKoodiUri] = curr.teksti || {};
    }

    return acc;
  }, {});

  return {
    tila,
    kieliversiot: kielivalinta,
    tarjoajat,
    information: {
      nimi,
      koulutus: {
        value: koulutusKoodiUri,
      },
      opintojenLaajuus: {
        value: opintojenLaajuusKoodiUri,
      },
      tutkintonimike: tutkintonimikeKoodiUrit.map(value => ({ value })),
      koulutusalat: koulutusalaKoodiUrit.map(value => ({ value })),
    },
    koulutustyyppi,
    lisatiedot: {
      osioKuvaukset,
      osiot,
    },
    description: {
      kuvaus,
      nimi: kuvauksenNimi,
    },
    julkinen,
    teemakuva,
  };
};

export default getFormValuesByKoulutus;
