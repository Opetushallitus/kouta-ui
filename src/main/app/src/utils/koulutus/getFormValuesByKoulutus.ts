import _fp from 'lodash/fp';

import { parseEditorState } from '#/src/components/LexicalEditorUI/utils';
import { KOULUTUSTYYPPI, MaaraTyyppi } from '#/src/constants';
import { KoulutusFormValues } from '#/src/types/koulutusTypes';
import { isNumeric, toSelectValue } from '#/src/utils';
import parseKoodiUri from '#/src/utils/koodi/parseKoodiUri';
import { isTutkintoonJohtavaKorkeakoulutus } from '#/src/utils/koulutus/isTutkintoonJohtavaKorkeakoulutus';

const koodiUriToKoodi = koodiUri => {
  return parseKoodiUri(koodiUri)?.koodiArvo;
};

// Backendiss채 koulutuskoodiUrit on tallennettu vain yhteen listaan. Frontissa on helpompi k채sitell채
// erikseen tapaus jossa koodiUreja voi olla vain yksi (muut kuin korkeakoulutukset) ja tapaus jossa
// niit채 voi olla monta (korkeakoulutus)
function getKoulutusKoodiUrit(
  koulutustyyppi: KOULUTUSTYYPPI,
  koulutusKoodiUrit?: Array<string>
): { koulutusKoodiUri?: string; korkeakoulutusKoodiUrit: Array<string> } {
  const isKorkeakoulu = isTutkintoonJohtavaKorkeakoulutus(koulutustyyppi);

  const firstElement = koulutusKoodiUrit?.[0];

  return {
    koulutusKoodiUri: isKorkeakoulu ? undefined : firstElement,
    korkeakoulutusKoodiUrit: isKorkeakoulu ? koulutusKoodiUrit || [] : [],
  };
}

export const getFormValuesByKoulutus = (koulutus): KoulutusFormValues => {
  const {
    kielivalinta = [],
    koulutustyyppi = '',
    tarjoajat = [],
    metadata = {},
    nimi = {},
    julkinen = false,
    esikatselu = false,
    tila,
    teemakuva,
    ePerusteId,
    sorakuvausId,
    externalId,
    organisaatioOid,
  } = koulutus;

  const { koulutusKoodiUri, korkeakoulutusKoodiUrit } = getKoulutusKoodiUrit(
    koulutustyyppi,
    koulutus.koulutuksetKoodiUri
  );

  const {
    lisatiedot = [],
    kuvaus = {},
    tutkinnonOsat = [],
    opintojenLaajuusyksikkoKoodiUri = '',
    opintojenLaajuusNumero,
    opintojenLaajuusNumeroMin,
    opintojenLaajuusNumeroMax,
    tutkintonimikeKoodiUrit = [],
    koulutusalaKoodiUrit = [],
    osaamisalaKoodiUri,
    linkkiEPerusteisiin = {},
    isAvoinKorkeakoulutus,
    tunniste,
    opinnonTyyppiKoodiUri,
    erikoistumiskoulutusKoodiUri = '',
    osaamismerkkiKoodiUri,
    luokittelutermit,
  } = metadata;

  return {
    organisaatioOid: toSelectValue(organisaatioOid),
    externalId,
    tila,
    kieliversiot: kielivalinta,
    tarjoajat: { tarjoajat },
    information: {
      nimi,
      eperuste: {
        value: ePerusteId,
      },
      koulutus: {
        value: koulutusKoodiUri,
      },
      korkeakoulutukset: korkeakoulutusKoodiUrit.map(value => ({ value })),
      opintojenLaajuusyksikko: {
        value: opintojenLaajuusyksikkoKoodiUri,
      },
      opintojenLaajuusNumero: isNumeric(opintojenLaajuusNumero)
        ? opintojenLaajuusNumero.toString()
        : '',
      laajuusNumeroTyyppi:
        opintojenLaajuusNumeroMin === opintojenLaajuusNumeroMax
          ? MaaraTyyppi.YKSI_ARVO
          : MaaraTyyppi.VAIHTELUVALI,
      opintojenLaajuusNumeroMin: opintojenLaajuusNumeroMin,
      opintojenLaajuusNumeroMax: opintojenLaajuusNumeroMax,
      tutkintonimike: tutkintonimikeKoodiUrit.map(value => ({ value })),
      koulutusalat: koulutusalaKoodiUrit.map(value => ({ value })),
      isAvoinKorkeakoulutus: Boolean(isAvoinKorkeakoulutus),
      tunniste,
      opinnonTyyppi: {
        value: opinnonTyyppiKoodiUri,
      },
      erikoistumiskoulutus: {
        value: erikoistumiskoulutusKoodiUri,
      },
      osaamismerkki: { value: osaamismerkkiKoodiUri },
      ...(![KOULUTUSTYYPPI.OSAAMISALA, KOULUTUSTYYPPI.TUTKINNON_OSA].includes(
        koulutustyyppi
      ) && {
        luokittelutermit: luokittelutermit?.map(
          (value: string): { label: string; value: string } => {
            return {
              label: value,
              value: value,
            };
          }
        ),
      }),
    },
    koulutustyyppi,
    lisatiedot: {
      osioKuvaukset: lisatiedot.reduce((acc, curr) => {
        if (curr.otsikkoKoodiUri) {
          acc[curr.otsikkoKoodiUri] = _fp.mapValues(
            parseEditorState,
            curr.teksti
          );
        }
        return acc;
      }, {}),
      osiot: lisatiedot
        .filter(({ otsikkoKoodiUri }) => Boolean(otsikkoKoodiUri))
        .map(({ otsikkoKoodiUri }) => ({ value: otsikkoKoodiUri })),
      ...([KOULUTUSTYYPPI.OSAAMISALA, KOULUTUSTYYPPI.TUTKINNON_OSA].includes(
        koulutustyyppi
      ) && {
        luokittelutermit: luokittelutermit?.map(
          (value: string): { label: string; value: string } => {
            return {
              label: value,
              value: value,
            };
          }
        ),
      }),
    },
    tutkinnonosat: {
      osat: _fp.values(
        _fp.reduce(
          (
            grouped,
            { ePerusteId, koulutusKoodiUri, tutkinnonosaId, tutkinnonosaViite }
          ) => ({
            ...grouped,
            [`${koulutusKoodiUri}_${ePerusteId}`]: {
              koulutus: { value: koulutusKoodiUri },
              eperuste: { value: ePerusteId },
              osat: [
                ...(grouped?.[`${koulutusKoodiUri}_${ePerusteId}`]?.osat || []),
                {
                  value: _fp.toString(tutkinnonosaId),
                  viite: _fp.toString(tutkinnonosaViite),
                },
              ],
            },
          }),
          {}
        )(tutkinnonOsat)
      ),
      nimi,
    },
    description: {
      kuvaus: _fp.mapValues(parseEditorState, kuvaus),
      linkkiEPerusteisiin,
    },
    esikatselu,
    julkinen,
    teemakuva,
    osaamisala: osaamisalaKoodiUri
      ? {
          osaamisala: {
            value: koodiUriToKoodi(osaamisalaKoodiUri),
          },
          eperuste: { value: ePerusteId },
          koulutus: { value: koulutusKoodiUri },
        }
      : null,
    soraKuvaus: sorakuvausId
      ? {
          value: sorakuvausId,
        }
      : null,
  };
};

export default getFormValuesByKoulutus;
