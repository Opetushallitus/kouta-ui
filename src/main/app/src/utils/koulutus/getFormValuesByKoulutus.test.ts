import getFormValuesByKoulutus from '#/src/utils/koulutus/getFormValuesByKoulutus';

test('getFormValuesByKoulutus returns correct form values given koulutus', () => {
  const values = getFormValuesByKoulutus({
    externalId: 'ext1',
    tila: 'tallennettu',
    julkinen: true,
    kielivalinta: ['fi', 'sv'],
    koulutuksetKoodiUri: ['koulutus_371101#1', 'koulutus_201000#1'],
    koulutustyyppi: 'yo',
    metadata: {
      kuvaus: {
        fi: 'Fi kuvaus',
        sv: 'Sv kuvaus',
      },
      lisatiedot: [
        {
          otsikkoKoodiUri: 'osio_1#1',
          teksti: {},
        },
        {
          otsikkoKoodiUri: 'osio_2#1',
          teksti: {
            fi: 'Fi kuvaus',
            sv: 'Sv kuvaus',
          },
        },
      ],
      tutkintonimikeKoodiUrit: ['nimike_1#1', 'nimike_2#1'],
      tyyppi: 'yo',
      koulutusalaKoodiUrit: ['koulutusala_1#1', 'koulutusala_2#1'],
    },
    nimi: {
      fi: 'Fi nimi',
      sv: 'Sv nimi',
    },
    tarjoajat: ['123.456.789'],
    sorakuvausId: '1234',
  });

  expect(values).toMatchSnapshot();
});

test('getFormValuesByKoulutus returns correct form values for ammatillinen koulutus', () => {
  const values = getFormValuesByKoulutus({
    tila: 'tallennettu',
    julkinen: true,
    kielivalinta: ['fi', 'sv'],
    koulutuksetKoodiUri: ['koulutus_371101#1'],
    koulutustyyppi: 'amm',
    metadata: {
      kuvaus: {
        fi: 'Fi kuvaus',
        sv: 'Sv kuvaus',
      },
      lisatiedot: [
        {
          otsikkoKoodiUri: 'osio_1#1',
          teksti: {},
        },
        {
          otsikkoKoodiUri: 'osio_2#1',
          teksti: {
            fi: 'Fi kuvaus',
            sv: 'Sv kuvaus',
          },
        },
      ],
      tutkintonimikeKoodiUrit: ['nimike_1#1', 'nimike_2#1'],
      tyyppi: 'amm',
      koulutusalaKoodiUrit: ['koulutusala_1#1', 'koulutusala_2#1'],
    },
    nimi: {
      fi: 'Fi nimi',
      sv: 'Sv nimi',
    },
    tarjoajat: ['123.456.789'],
  });

  expect(values).toMatchSnapshot();
});

test('getFormValuesByKoulutus returns correct form values for erikoistumiskoulutus', () => {
  const values = getFormValuesByKoulutus({
    koulutustyyppi: 'erikoistumiskoulutus',
    metadata: {
      kuvaus: {
        fi: 'Fi kuvaus',
        sv: 'Sv kuvaus',
      },
      erikoistumiskoulutusKoodiUri: 'erikoistumiskoulutukset_001#2',
      opintojenLaajuusyksikko: 'opintojenlaajuusyksikko_2#1',
      opintojenLaajuusNumeroMax: 10,
      opintojenLaajuusNumeroMin: 5,
      koulutusalaKoodiUrit: ['koulutusala_1#1', 'koulutusala_2#1'],
    },
  });

  expect(values).toMatchSnapshot();
});
