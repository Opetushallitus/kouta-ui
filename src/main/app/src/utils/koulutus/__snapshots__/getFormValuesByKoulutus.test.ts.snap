// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getFormValuesByKoulutus returns correct form values for ammatillinen koulutus 1`] = `
Object {
  "description": Object {
    "kuvaus": Object {
      "fi": "<p>Fi kuvaus</p>",
      "sv": "<p>Sv kuvaus</p>",
    },
    "linkkiEPerusteisiin": Object {},
  },
  "esikatselu": false,
  "externalId": undefined,
  "information": Object {
    "eperuste": Object {
      "value": undefined,
    },
    "korkeakoulutukset": Array [],
    "koulutus": Object {
      "value": "koulutus_371101#1",
    },
    "koulutusalat": Array [
      Object {
        "value": "koulutusala_1#1",
      },
      Object {
        "value": "koulutusala_2#1",
      },
    ],
    "nimi": Object {
      "fi": "Fi nimi",
      "sv": "Sv nimi",
    },
    "opintojenLaajuus": Object {
      "value": "laajuus_1#1",
    },
    "opintojenLaajuusRange": "",
    "opintojenLaajuusnumero": "",
    "opintojenLaajuusyksikko": Object {
      "value": "",
    },
    "tutkintonimike": Array [
      Object {
        "value": "nimike_1#1",
      },
      Object {
        "value": "nimike_2#1",
      },
    ],
  },
  "julkinen": true,
  "kieliversiot": Array [
    "fi",
    "sv",
  ],
  "koulutustyyppi": "amm",
  "lisatiedot": Object {
    "osioKuvaukset": Object {
      "osio_1#1": Object {},
      "osio_2#1": Object {
        "fi": "<p>Fi kuvaus</p>",
        "sv": "<p>Sv kuvaus</p>",
      },
    },
    "osiot": Array [
      Object {
        "value": "osio_1#1",
      },
      Object {
        "value": "osio_2#1",
      },
    ],
  },
  "organisaatioOid": undefined,
  "osaamisala": null,
  "soraKuvaus": null,
  "tarjoajat": Object {
    "tarjoajat": Array [
      "123.456.789",
    ],
  },
  "teemakuva": undefined,
  "tila": "tallennettu",
  "tutkinnonosat": Object {
    "nimi": Object {
      "fi": "Fi nimi",
      "sv": "Sv nimi",
    },
    "osat": Array [],
  },
}
`;

exports[`getFormValuesByKoulutus returns correct form values given koulutus 1`] = `
Object {
  "description": Object {
    "kuvaus": Object {
      "fi": "<p>Fi kuvaus</p>",
      "sv": "<p>Sv kuvaus</p>",
    },
    "linkkiEPerusteisiin": Object {},
  },
  "esikatselu": false,
  "externalId": "ext1",
  "information": Object {
    "eperuste": Object {
      "value": undefined,
    },
    "korkeakoulutukset": Array [
      Object {
        "value": "koulutus_371101#1",
      },
      Object {
        "value": "koulutus_201000#1",
      },
    ],
    "koulutus": Object {
      "value": undefined,
    },
    "koulutusalat": Array [
      Object {
        "value": "koulutusala_1#1",
      },
      Object {
        "value": "koulutusala_2#1",
      },
    ],
    "nimi": Object {
      "fi": "Fi nimi",
      "sv": "Sv nimi",
    },
    "opintojenLaajuus": Object {
      "value": "laajuus_1#1",
    },
    "opintojenLaajuusRange": "",
    "opintojenLaajuusnumero": "",
    "opintojenLaajuusyksikko": Object {
      "value": "",
    },
    "tutkintonimike": Array [
      Object {
        "value": "nimike_1#1",
      },
      Object {
        "value": "nimike_2#1",
      },
    ],
  },
  "julkinen": true,
  "kieliversiot": Array [
    "fi",
    "sv",
  ],
  "koulutustyyppi": "yo",
  "lisatiedot": Object {
    "osioKuvaukset": Object {
      "osio_1#1": Object {},
      "osio_2#1": Object {
        "fi": "<p>Fi kuvaus</p>",
        "sv": "<p>Sv kuvaus</p>",
      },
    },
    "osiot": Array [
      Object {
        "value": "osio_1#1",
      },
      Object {
        "value": "osio_2#1",
      },
    ],
  },
  "organisaatioOid": undefined,
  "osaamisala": null,
  "soraKuvaus": Object {
    "value": "1234",
  },
  "tarjoajat": Object {
    "tarjoajat": Array [
      "123.456.789",
    ],
  },
  "teemakuva": undefined,
  "tila": "tallennettu",
  "tutkinnonosat": Object {
    "nimi": Object {
      "fi": "Fi nimi",
      "sv": "Sv nimi",
    },
    "osat": Array [],
  },
}
`;
